Sources/DataConnectError.swift

```diff
@@ -14,27 +14,239 @@
 
 import Foundation
 
-/// Represents an error returned by the DataConnect service
+// MARK: Base Error Definitions
+
+/// Protocol representing an error returned by the DataConnect service
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+public protocol DataConnectError: Error, CustomDebugStringConvertible, CustomStringConvertible {
+  var message: String? { get }
+  var cause: Error? { get }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+public extension DataConnectError {
+  var debugDescription: String {
+    return "{\(Self.self), message: \(message ?? "nil"), cause: \(String(describing: cause))}"
+  }
+
+  var description: String {
+    return debugDescription
+  }
+}
+
+/// Type erased DataConnectError
 @available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
-public enum DataConnectError: Error {
-  /// no firebase app specified. configure not complete
-  case appNotConfigured
+public struct AnyDataConnectError: DataConnectError {
+  private let dataConnectError: DataConnectError
+
+  init<E: DataConnectError>(dataConnectError: E) {
+    self.dataConnectError = dataConnectError
+  }
+
+  public var message: String? {
+    return dataConnectError.message
+  }
+
+  public var cause: (any Error)? {
+    return dataConnectError.cause
+  }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+/// Represents an error domain which can have more granular error codes
+public protocol DataConnectDomainError: DataConnectError {
+  associatedtype ErrorCode: DataConnectErrorCode
+
+  var code: ErrorCode { get }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+public extension DataConnectDomainError {
+  var debugDescription: String {
+    return "{\(Self.self), code: \(code), message: \(message ?? "nil"), cause: \(String(describing: cause))}"
+  }
+
+  var description: String {
+    return debugDescription
+  }
+}
+
+/// Error code within an error domain
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+public protocol DataConnectErrorCode: CustomStringConvertible, Equatable, Sendable, CaseIterable {}
+
+// MARK: Data Connect Initialization Errors
+
+/// Error initializing Data Connect
+public struct DataConnectInitError: DataConnectDomainError {
+  public struct Code: DataConnectErrorCode {
+    private let code: String
+    private init(_ code: String) { self.code = code }
+
+    public static let appNotConfigured = Code("appNotConfigured")
+    public static let grpcNotConfigured = Code("grpcNotConfigured")
+
+    public static var allCases: [DataConnectInitError.Code] {
+      return [appNotConfigured, grpcNotConfigured]
+    }
+
+    public var description: String { return code }
+  }
+
+  public let code: Code
+
+  public private(set) var message: String?
+
+  public private(set) var cause: Error?
```

Comment:
looks like these can be constants
```suggestion
  public let message: String?

  public let cause: Error?
```